EIPS/eip-7002.md

```diff
@@ -0,0 +1,380 @@
+num_exits_in_queue = queue_tail_index - queue_head_index
+num_exits_to_dequeue = min(num_exits_in_queue, MAX_EXITS_PER_BLOCK)
+
+# Retrieve exits from the queue
+expected_exits = []
+for i in range(num_exits_to_dequeue):
+    queue_storage_slot = EXIT_MESSAGE_QUEUE_STORAGE_OFFSET + (queue_head_index + i) * 2
+    source_address = address(SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot)[0:20]
+    validator_pubkey = (
+        SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1) + SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1)
+        + SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1) + SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1)[0:16]
+    )
+    exit = Exit(
+        source_address=Bytes20(source_address),
+        validator_pubkey=Bytes48(validator_pubkey),
+    )
+    expected_exits.append(exit)
+
+# Compare retrieved exits to the list in the block body
+assert block.body.exits == expected_exits
+```
+
+A block that does not satisfy the above conditions **MUST** be deemed invalid.
+
+#### Block processing
+
+##### Per-block precompile storage calculations
+
+At the end of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. after processing all transactions and after performing the block body exit validations):
+
+* The exit precompile's exit queue is updated based on exits dequeued and the exit queue head/tail are reset if the queue has been cleared (`update_exit_queue()`)
+* The exit precompile’s excess exits are updated based on usage in the current block (`update_excess_exits()`)
+* The exit precompile's exit count is reset to 0 (`reset_exit_count()`)
+
+Specifically, the functionality is defined in pseudocode as the function `update_exit_precompile()`:
+
+```python
+###################
+# Public function #
+###################
+
+def update_exit_precompile():
+    update_exit_queue()
+    update_excess_exits()
+    reset_exit_count()
+    
+###########
+# Helpers #
+###########
+
+def update_exit_queue():
+    queue_head_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT)
+    queue_tail_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_TAIL_STORAGE_SLOT)
+    
+    num_exits_in_queue = queue_tail_index - queue_head_index
+    num_exits_dequeued = min(num_exits_in_queue, MAX_EXITS_PER_BLOCK)
+    new_queue_head_index = queue_head_index + num_exits_dequeued
+    if new_queue_head_index == queue_tail_index:
+        # Queue is empty, reset queue pointers
+        sstore(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT, 0)
+        sstore(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_TAIL_STORAGE_SLOT, 0)
+    else:
+        sstore(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT, new_queue_head_index)
+
+def update_excess_exits():
+    previous_excess_exits = sload(EXIT_PRECOMPILE_ADDRESS, EXCESS_EXITS_STORAGE_SLOT)
+    exit_count = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_COUNT_STORAGE_SLOT)
```

Comment:
it seems like we could just pass in this data `len(block.body.exits)` as part of the precompile update routines and skip the machinery to update/clear the exit count in the storage (and it would make the required additional protocol state a bit smaller)